File: 01/rust/src/main.rs
use itertools::Itertools;
use std::num::ParseIntError;

fn main() -> anyhow::Result<()> {
    let input = include_str!("../../input.txt")
        .lines()
        .map(|line| line.parse::<usize>())
        .collect::<Result<Vec<usize>, ParseIntError>>()?;
    let part1 = solve_part1(&input);
    println!("part 1: {}", part1);
    let part2 = solve_part2(&input);
    println!("part 2: {}", part2);
    Ok(())
}

fn solve_part1(input: &[usize]) -> usize {
    input
        .windows(2)
        .filter(|window| window[0] < window[1])
        .count()
}

fn solve_part2(input: &[usize]) -> usize {
    input
        .windows(3)
        .map(|window| window.iter().sum::<usize>())
        .tuple_windows()
        .filter(|(prev, next)| prev < next)
        .count()
}

File: 01/python/main.py
def solve_part1(depths):
    return sum(prev < curr for prev, curr in zip(depths, depths[1:]))


def solve_part2(depths):
    window_sums = [a+b+c for a, b, c in zip(depths, depths[1:], depths[2:])]
    return solve_part1(window_sums)


with open("../input.txt") as f:
    depths = [int(line) for line in f]
    part1 = solve_part1(depths)
    print(part1)
    part2 = solve_part2(depths)
    print(part2)
